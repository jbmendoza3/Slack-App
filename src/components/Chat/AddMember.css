
.add-member-modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
  }
  
  .modal-content {
    background-color: rgb(5, 5, 5);
    padding: 20px;
    border-radius: 8px;
    width: 350px;
    text-align: center;
    border-color: antiquewhite;
    font-size: 20px;
  }
  
  .modal-buttons {
    margin-top: 10px;
    display: flex;
    justify-content: space-between;
  }
  
  .error-message {
    color: red;
  }
  