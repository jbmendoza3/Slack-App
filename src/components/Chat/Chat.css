.chat-container {
    padding: 20px;
    padding-top: 50px;
    background-color: rgb(5, 5, 5);
    border-radius: 8px;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
    margin-left: 23rem;
    height: 100%;
    position: inherit;
    z-index: -1;
    width: 73%;
    color: whitesmoke;
}
  
.chat-box {
    height: 750px;
    overflow-y: scroll;
    border: 1px solid #ddd;
    padding: 10px;
    background-color: rgb(15, 15, 15);
    margin-bottom: 10px;
    color: white;
    font-size: 17px;
    text-align: left;
}

.chat-box span {
    font-size: 12px;
    font-weight: 700;
    padding-top: 5px;
}

  
.chat-input {
    display: flex;
    flex-direction: center;
    align-items: center;
}
  
.chat-input input {
    flex: 1;
    padding: 10px;
    margin-top: 15px;
    border: 1px solid #ddd;
    border-radius: 4px;
    margin-right: 10px;
}

.chat-input input[type="text"] {
    background-color: rgb(68, 68, 68);
    width: 50rem;
    color: whitesmoke;
    border-radius: 0;
    text-align: left;
}
  
.chat-input button {
    padding: 10px 20px;
    height: 45px;
    border: none;
    background-color: #007bff;
    color: #fff;
    border-radius: 4px;
    cursor: pointer;
}
  
.chat-input button:hover {
    background-color: #0056b3;
}
  
.chat-header {
    display: flex;
    justify-content: space-between;
    text-align: center;
    font-size: 20px;
    font-weight: 900;
}

.headerLeft {
    display: flex;
    flex-direction: row;
    justify-content: center;
    padding: 10px 20px;
    background-color: rgb(5, 5, 5);
}

.headerRight {
    display: flex;
    justify-content: space-between;
}

.add-member-button {
    padding: 5px 7px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    background-color: rgb(5, 5, 5);
    color: white;
}

.userlist-button {
    padding: 5px 7px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    background-color: rgb(5, 5, 5);
    color: white;
}

.timestamp {
    color: rgb(68, 68, 68);
}

.message-body-content {
    font-size: 18px;
    padding-top: 3px;
}
