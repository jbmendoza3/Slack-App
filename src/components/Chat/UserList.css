.user-list-modal {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: transparent;
    display: flex;
    justify-content: center;
    align-items: center;
    color: whitesmoke;
  }
  
  .modal-content {
    background-color: rgb(51, 51, 51);
    padding: 20px;
    border-radius: 8px;
    width: 300px;
    text-align: center;
    font-size: 20px;
  }
  
  ul {
    list-style-type: none;
    padding: 0;
  }
  
  li {
    margin: 10px 0;
  }
  