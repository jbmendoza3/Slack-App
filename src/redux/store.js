import { configureStore } from '@reduxjs/toolkit';
import chatReducer from './chatSlice';

const store = configureStore({
  reducer: {
    chat: chatReducer
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      thunk: true
    })
});

export default store;
