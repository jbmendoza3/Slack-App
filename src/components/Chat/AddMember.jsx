import React, { useState } from 'react';
import AddMemberService from '../../services/AddMemberService';
import './AddMember.css';

const AddMember = ({ channelId, onClose }) => {
  const [userId, setUserId] = useState('');
  const user = JSON.parse(localStorage.getItem('user'));

  const handleAddMember = async () => {
    try {

      console.log("User:", user);
      console.log("UserId (input)", userId);
      console.log("ChannelId", channelId);

      const response = await AddMemberService.addMember(user, channelId);
      console.log("Response",response);
      if (response) {
        alert('Member added successfully');
        onClose();
      } 
    } catch (error) {
      alert('Error adding member');
      console.log("Error details", error);
    }
  };

  return (
    <div className="add-member-modal">
      <div className="modal-content">
        <h2>Add Member</h2>
        <input
          type="text"
          value={userId}
          onChange={(e) => setUserId(e.target.value)}
          placeholder="Enter user ID"
        />
        <div className="modal-buttons">
          <button onClick={handleAddMember}>Add Member</button>
          <button onClick={onClose}>Cancel</button>
        </div>
      </div>
    </div>
  );
};

export default AddMember;
