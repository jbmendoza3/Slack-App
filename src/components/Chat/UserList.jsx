import React, { useEffect, useState } from 'react';
import UserService from '../../services/UserService';
import './UserList.css';

const UserList = ({ channelId, onClose }) => {
  const [users, setUsers] = useState([]);
  const user = JSON.parse(localStorage.getItem('user'));

  useEffect(() => {
    const fetchUsers = async () => {
        if(users.length <= 0) {
            try {
                const usersInChannel = await UserService.getUsersInChannel(user, channelId);
                setUsers(usersInChannel);
              } catch (error) {
                alert('Error fetching user list');
            }
        }
        console.log(users)
    };

    fetchUsers();
  }, [channelId, user, users]);

  return (
    <div className="user-list-modal">
      <div className="modal-content">
        <h2>Channel Users</h2>
        <ul>
          {users && users.map((person) => (
            <li key={person["user_id"]}>({person["user_id"]})</li>
          ))}
        </ul>
        <button onClick={onClose}>Close</button>
      </div>
    </div>
  );
};

export default UserList;
