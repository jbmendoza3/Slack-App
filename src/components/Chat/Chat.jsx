import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { fetchMessages } from "../../redux/chatSlice";
import "./Chat.css";
import ChatService from "../../services/ChatService";
import PeopleAltIcon from '@mui/icons-material/PeopleAlt';
import MailIcon from '@mui/icons-material/Mail';
import AddMember from './AddMember';
import UserList from './UserList';
import AddIcon from '@mui/icons-material/Add';


const Chat = () => {
   const dispatch = useDispatch();
   const { selectedChannel, messages } = useSelector((state) => state.chat);
   const user = JSON.parse(localStorage.getItem("user"));
   const [newMessage, setNewMessage] = useState("");
   const [chatFlag, setChatFlag] = useState(false);
   const [isAddMemberOpen, setIsAddMemberOpen] = useState(false);
   const [listMemberOpen, setlistMemberOpen] = useState(false);
   

   useEffect(() => {
      const getMessages = () => {
         if (selectedChannel) {
            console.log(selectedChannel.id);
            dispatch(fetchMessages(user, selectedChannel.id));
         }
      };

      if (chatFlag === false) {
         getMessages();
         setChatFlag(true);
      }
   }, [chatFlag, user, selectedChannel, dispatch]);

   const handleSendMessage = async () => {
      if (!newMessage.trim()) return;
      if (selectedChannel) {
         try {
            await ChatService.sendMessage(user, newMessage, selectedChannel.id);
            dispatch(fetchMessages(selectedChannel.id));
            setNewMessage("");
         } catch (error) {
            alert("Error sending message");
         }
      }
   };

   const handleToggleAddMember = () => {
      setIsAddMemberOpen(!isAddMemberOpen);
   };

   const handleToggleListMembers = () => {
      setlistMemberOpen(!listMemberOpen);

   };

   const formatTimestamp = (timestamp) => {
      const date = new Date(timestamp);
      return `${date.toLocaleDateString()} ${date.toLocaleTimeString()}`;
   };

   return (
      <div className="chat-container">
         {selectedChannel && (
          <>
            <div className="chat-header">
               <div className="headerLeft">
                  <MailIcon />
                  {selectedChannel ? selectedChannel.name : "Select a channel"}
               </div>
            <div className="headerRight">
                  <button className="add-member-button" onClick={handleToggleAddMember}>
                     <AddIcon />
                  </button>
            
                  <button className="userlist-button" onClick={handleToggleListMembers}>
                     <PeopleAltIcon />
                  </button>
            </div>
             </div>
         <div className="chat-box">
            {messages.data &&
               messages.data.map((msg, index) => (
                     <div key={index} className={user.uid === msg.sender.uid ? "mapMessages-user-container" : "mapMessages-receiver-container"}>
                         <span className="currentUser">
                            {msg.sender.uid}
                         </span>
                        <p className="message-body-content">{msg.body}</p>
                        <span className="timestamp">{formatTimestamp(msg["created_at"])}</span>
                     </div>
               ))}
         </div>
         <div className="chat-input">
            <input
               type="text"
               value={newMessage}
               onChange={(e) => setNewMessage(e.target.value)}
               placeholder="Type a message..."
            />
            <button onClick={handleSendMessage}>Send</button>
         </div>
         {listMemberOpen && (
            <UserList channelId={selectedChannel.id} onClose={handleToggleListMembers} />
         )}
         {isAddMemberOpen && (
            <AddMember channelId={selectedChannel.id} onClose={handleToggleAddMember} />
         )}
         </>
         )}
         
      </div>
   );
};

export default Chat;
