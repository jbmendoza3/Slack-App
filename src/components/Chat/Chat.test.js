import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Chat from './Chat';
import store from '../../redux/store'; // assuming your store setup is in this file
import renderer from 'react-test-renderer';

describe('<Chat />', () => {
  test('renders chat component', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Chat />
        </Router>
      </Provider>
    );

    await waitFor(() => {
      expect(store.getState().chat.selectedChannel).toBeDefined();
    });

    // const heading = screen.getByRole('heading');
    // expect(heading).toBeInTheDocument();
    // expect(heading).toHaveTextContent(/Select a channel/i);
  });
});

test('snapshot test for Chat', async () => {
  let chat;
  await renderer.act(async () => {
    chat = renderer.create(
      <Provider store={store}>
        <Router>
          <Chat />
        </Router>
      </Provider>
    );
  });

  expect(chat.toJSON()).toMatchSnapshot();
});
