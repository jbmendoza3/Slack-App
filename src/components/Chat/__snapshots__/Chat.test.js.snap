// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> snapshot test for Chat 1`] = `
<div
  className="chat-container"
/>
`;

exports[`snapshot test for Chat 1`] = `
<div
  className="chat-container"
/>
`;
